ops/testing.py

```diff
@@ -2586,6 +2606,118 @@ class NonAbsolutePathError(Exception):
     """
 
 
+class ContainerFilesystem(Protocol):
+    """The subset of container operations that are filesystem-related.
+
+    Don't instantiate this directly; use :meth:`Harness.get_container_filesystem`.
+    """
+
+    @abstractmethod
```

Comment:
Per [the Protocol PEP 544](https://peps.python.org/pep-0544/), it's best if abstract methods `raise NotImplementedError` in their body. It probably doesn't matter much here as we're always using the proxy, but just in case we accidentally instantiate/subclass this, might as well do that on all these methods.

Reply:
Updated, although I have to complete some type hints in the `ops.model.Container` for type checks to pass.